### PR TITLE
Alter ISISReflectometryCalibration algorithm

### PR DESCRIPTION
**Pull request by rbauststfc**
**Description of work**

**Purpose of work**

A calibration step will be required when the new 2D detector starts being used on INTER in June. This calibration algorithm will also be useful for POLREF eventually.

**Summary of work**

I've re-written algorithm `ReflectometryISISCalibration` to implement the approach described in the referenced issue. Our Reflectometry scientists have a need to move individual detector pixels rather than moving the entire bank, as is sometimes done in other calibrations. They measure theta offsets for each pixel, which are then used to find the new detector pixel position. Existing algorithm `SpecularReflectionPositionCorrect` can calculate and apply the new positions, but internally uses method `MoveInstrumentComponent` to do this, which prevents moving individual pixels for rectangular (and other fixed) detectors, and affects INTER. After some discussion, we concluded that there wasn't a good science reason for this obstacle and so I have added a boolean parameter to `SpecularReflectionPositionCorrect` and `MoveInstrumentComponent` that allows the user to choose to force individual pixel moves for fixed detectors. The new parameter will default to preventing these moves to maintain backwards compatibility.

**Further detail of work**

The following tasks are still outstanding for this issue, but I would like to complete them a separate PR because of the need to get this merged in and tested by our users in time for the 6.7 release:

- Documentation updates - I need to add documentation for the `ReflectometryISISCalibration` algorithm. I will also review documentation for the `SpecularReflectionPositionCorrect` and `MoveInstrumentComponent` algorithms to see if anything needs updating after adding the new parameter. Finally, the docs for algorithm `ApplyCalibration` incorrectly state that it will not move rectangular detectors. This does not appear to be the case and so this should be removed.

- It would be good to add additional unit tests to cover more of the calibration file parsing logic. I would like to take a bit of time to see if there is a way of doing this that doesn't require lots of different test files to be provided.

- I need to add a unit test to cover the case where the algorithm is moving a detector pixel that has a negative theta value to start with.

- Issue #34906 still exists and would prevent POLREF using the algorithm via the GUI. Fixing this is not required for INTER, though, which is the current priority.

- If individual detector pixels are moved for a rectangular detector then Mantid currently prevents saving the calibrated workspace as a Nexus file. At the moment Reflectometry do not really use this functionality, but we will still need to find a solution to this eventually. This will be done under a separate issue - #35498.

**To test:**

1) Download this test INTER calibration file and remove the .txt extension (added just to be able to attach to GitHub): 
[INTER_calibration_full.dat.txt](https://github.com/mantidproject/mantid/files/11371051/INTER_calibration_full.dat.txt)
1) Open the ISIS Reflectometry GUI
1) In the runs table, enter 61135 under the Run column and 0.5 under Angle.
1) On the Experiment Settings tab, set the AnalysisMode to MultiDetectorAnalysis.
1) On the Instrument Settings tab, set the CalibrationFile path by browsing to the location of the downloaded calibration file.
1) Back on the Runs tab, click the Process button.
1) The reduction should complete successfully. If you right click on the TOF output workspace and choose Show Detectors you can view the theta values of each calibrated detector.
1) Use Load from the Workspaces pane in the main window to load the original INTER61135 data from the archive. Check the original theta values using Show Detector. Compare these to the theta values from the reduced, calibrated workspace. The theta values in the calibrated workspaces should be equal to original value + theta offset (taken from the test calibration file).
1) Check that changing the format of the test calibration file gives either an appropriate error message (i.e. providing invalid columns), or doesn't prevent the calibration from succeeding.

Refs #35352. 

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: docs/source/release/v6.7.0/Reflectometry/New_features/35494.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+- The ReflectometryISISCalibration algorithm now calculates and performs individual detector pixel corrections from an input file containing detector IDs and theta offsets.
```
Comment (ConorMFinn) [inline docs/source/release/v6.7.0/Reflectometry/New_features/35494.rst:1]:
```suggestion
- The :ref:`algm-ReflectometryISISCalibration` algorithm now calculates and performs individual detector pixel corrections from an input file containing detector IDs and theta offsets.
```

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py [hunk 2]
```diff
@@ -16,14 +14,18 @@
...
 import math
 
 
 class ReflectometryISISCalibration(DataProcessorAlgorithm):
     _WORKSPACE = "InputWorkspace"
     _CALIBRATION_FILE = "CalibrationFile"
     _OUTPUT_WORKSPACE = "OutputWorkspace"
-    _DEBUG = "Debug"
+
+    _DET_ID_LABEL = "detectorid"
```
Comment (ConorMFinn) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py:26]: Is there a reason this isn't `detector_id`? Just to make it the same as the theta label

**Review by cailafinn — CHANGES_REQUESTED** [review]
Changes are working well, just a few code/usability notes:

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py [hunk 3]
```diff
@@ -35,152 +37,141 @@ def name(self):
...
+            calibration_data = self._parse_calibration_file(self._calibration_filepath)
+        except FileNotFoundError:
+            raise FileNotFoundError("Calibration file path cannot be found")
 
-        # Apply calibration to workspace
-        output_ws = CloneWorkspace(InputWorkspace=ws)
-        ApplyCalibration(Workspace=output_ws, CalibrationTable=calib_table)
-        self.setProperty(self._OUTPUT_WORKSPACE, output_ws)
-        AnalysisDataService.remove("output_ws")
-        if self.getProperty(self._DEBUG).isDefault:
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py:71]: The debug property was useful originally when we were using the ApplyCalibration algorithm, as we first created a TableWorkspace with all the values to be applied and then passed that into the algorithm. In debug mode we would output that TableWorkspace to the ADS. However now we don't build a table, each pixel is moved individually so there is no debug table to output. If users need to check the new TwoTheta values that have been calculated and applied to each detector they can now see that via the workspace history.

File: docs/source/release/v6.7.0/Reflectometry/New_features/35494.rst (hunk 1 above)
Comment (rbauststfc) [reply docs/source/release/v6.7.0/Reflectometry/New_features/35494.rst:1]: Will that work OK if the documentation doesn't exist yet? I'd deliberately left it out for now assuming I would need to add it in a later PR that adds the docs, but let me know if I'm misunderstanding how this works.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py [hunk 4]
```diff
@@ -35,152 +37,141 @@ def name(self):
...
+                    # The labels should be the first row in the file that doesn't begin with a #
+                    self._check_file_column_labels(entries)
+                    labels_checked = True
+                    continue
+
+                try:
+                    scanned_theta_offsets[int(entries[self._det_id_col_idx])] = float(entries[self._theta_offset_col_idx])
+                except ValueError:
+                    raise ValueError(
+                        "Invalid data in calibration file - detector ids should be integers and theta offsets should be floats"
```
[the remark this answers is not in the document]
Comment (rbauststfc) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py:119]: Yes, that's a good thought. I originally added this in because I had an issue when I put the labels in the wrong way round, but it could potentially just affect a single row.

[the remark this answers is not in the document]
Comment (rbauststfc) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py:119]: I'm going to print out the entry as part of the error message and see how the users get on with that for now, given that this error can come about a couple of ways. My understanding is that the users are planning to produce their calibration file via a script, so hopefully they should be unlikely to encounter this in practice.

**Comment by rbauststfc** [comment]
I'm putting this back into draft for now to consider some performance questions.

**Comment by rbauststfc** [comment]
I've done some testing to check the performance of the algorithm when calibrating a detector with 2048 pixels (to match the size of the new INTER detector). I found the following:

- Having a large history makes re-loading a saved calibration workspace very slow. As a result I have disabled recording the history for the calls to `SpecularReflectionPositionCorrect`.
- Re-loading a workspace with changes to 2048 individual pixel positions is only slower by a fraction of a second (on my machine) than a workspace without these changes, so I don't think we need to be concerned about this.
- The algorithm originally used `SpectrumInfo::signedTwoTheta` to get current detector two theta values, however this took a workspace index and getting the workspace index from a detector ID was slowing things down significantly with 2048 pixels (the whole calibration would take approx. 12 seconds to complete on my machine). Exposing `DetectorInfo::signedTwoTheta` in Python and using this instead reduced the calibration time to approx. 1.5 seconds for 2048 pixels.

I am investigating whether providing our own implementation of the theta to cartesian co-ordinate conversion that's within `SpecularReflectionPositionCorrect` and then using `ApplyCalibration` would give us better performance, history recording and debug information. However, because of the time constraint on this work, I think this first version would be a good start to give the scientists something to test and confirm the approach is working for them with the new detector before investing more development time. It would also be good to get this first version onto IDAaaS to do some performance testing there, as that's the platform we expect our scientists to be using most of the time.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rbauststfc) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py:71]: Since my latest changes it won't be possible to see the history any more, but users can check the theta values in the detector table if they want to inspect the output of the calibration step. This should give sufficient visibility - the other reason for having the explicit debug option previously was that when we were changing detector positions instead of two theta values it would have been harder for them to see the effect of the calibration.

**Review by cailafinn — COMMENTED** [review]
Looks good to me now. Working well! Nice work 😄 

(Approving now in light of jenkins being in shutdown thanks to some GitHub issues.)

**Review by cailafinn — APPROVED** [review]

**Comment by sf1919** [comment]
@mantidproject/gatekeepers and just in case Github goes down again in the next few days I'm adding this screenshot to show all tests passed and it's been approved
Screenshot (266)
